### ISSUE TITLE
[12.0] delegated inheritance fields' values are ignored on write during constrains check

### ISSUE THREAD

**Issue by oerp-odoo** [odoo/odoo#32919]
**Impacted versions:**
12.0
**Steps to reproduce:**
Have delegated inheritance models, like `product.product` -> `product.template`. Execute `write`, so it would trigger constraint call (with `api.constrains`). For example in `product.product`, there is constraint method called `_check_attribute_value_ids`. Also during that trigger, in `write` method also change field value that is only defined on `product.template`, but not redefined on `product.product`. For example field called `rental`. Also it does not matter if `rental` would be included in `api.constrains` as dependency (it behaves the same way).

Code example:
```python
product = self.env.ref('product.product_product_4d')
attr = product.attribute_value_ids[0]
attr2 = product.attribute_value_ids[1]
p = self.env['product.product'].create(
    {
        'name': 'T1',
        'attribute_value_ids': [(6, 0, [attr.id])],
        'rental': True
    })
p.write({
    'attribute_value_ids': [(6, 0, [attr2.id])],
    'rental': False
})
```
**Log `rental` field value inside constraint method body.**

**Current behavior:**
It will log delegated field's old value (as in this example `rental` field) inside constraint, though it will write that value correctly. In other sense, constraint will not see that such field is being updated, meaning it bypasses constraint.

Inside `_check_attribute_value_ids` method:
  * during create, `rental` value will be printed `True`. OK
  * during write, `rental` value will be printed `True`. Failed.

As can be seen in this example, on create, it works as expected, it does output the value as it was set on create. But during write, it does not (in this example, I try to set `rental` to `False`, but it still prints `True`).

**Expected behavior:**
It should log rental field value as it was set on `write` call.

Other notes. This probably might cause issues for other methods like compute methods (with `api.depends` decorator)?
I also checked source code in `odoo/odoo/models.py` and found this change https://github.com/odoo/odoo/commit/be0f24a96bcd6d66448f422d21ab126297885bd9

Specifically look at this block of code (3013 - 3024 lines):  `# update parent records (after possibly updating parent fields)`

It uses raw SQL to update parent records, possibly ignoring constraint update for such fields? I don't see this commit on odoo 11 and there is no such issue as this one. Maybe this change has something to do with it?

**Video/Screenshot link (optional):**
[delegated-inheritance-bug.ogv.zip](https://github.com/odoo/odoo/files/3112256/delegated-inheritance-bug.ogv.zip)

P.S. In video, no standard code was modified, only `_check_attribute_value_ids` method was injected with logging to output tested values.

**Comment by Julien00859** [comment]
I might not get it, the value is effectively updated by the write, what's wrong ?

	Python 3.6.5 (default, Apr  1 2018, 05:46:30) 
	[GCC 7.3.0] on linux
	Type "help", "copyright", "credits" or "license" for more information.
	(Console)
	>>> pro = self.env.ref('product.product_product_4d')
	>>> p = self.env['product.product'].create({'name': 'T1', 'attribute_value_ids': [(6, 0, [pro.attribute_value_ids[0].id])], 'rental': True})
	>>> p.rental
	True
	>>> p.write({'attribute_value_ids': [(6, 0, [pro.attribute_value_ids[1].id])], 'rental': False})
	True
	>>> p.rental
	False

**Comment by oerp-odoo** [comment]
@Julien00859 yes you didn't get it. Yes write values are updated, though those values are ignored during constraint methods call (as in my example during `_check_attribute_value_ids` call) . With those mentioned values (like rental), it works as if there is no constraint, or like plain SQL update.

**Comment by oerp-odoo** [comment]
@Julien00859 @rco-odoo This fix does not work correctly. I just had a bit different case, when it failed. Yes it now passes parent field new values on constraint. But the problem is now that if constraint consists only of parent fields, constraint is never triggered. Meaning it can still bypass constraint. It only triggers if `constrains` decorator has at least one field that is not parent (and such field value is changed alongside parent field).

Also unittests that check if this fix is correct - passes , is because it was not testing how it behaves with parent fields only (field `val1` is parent field, and `val2` is model's field which has constraint defined, so thats why this bug slips through that unittest).

For example (in `product.product`):

This one would work in some cases (because we have `default_code` field, that is also defined in product.product).

```python
@api.one
@api.constrains('rental', 'default_code')
def _check_rental(self):
    print(self.rental, self.default_code)
```
# triggers constraint.
`product.write({'rental': True, 'default_code': 'some_new_code'})`
# does not trigger constraint.
`product.rental = False`

This does not trigger constraint at all.

```python
@api.one
@api.constrains('rental')
def _check_rental(self):
    print(self.rental)
```

# wont trigger constraint.
`product.rental = True`

**Comment by Julien00859** [comment]
I tried your behaviour too during my fix and was doubtfully why it was not working.

The reason is that when you write fields on the parent, only constraints of the parent are triggered. In current Odoo constraints in the child model solely checking parent fields are never triggered as the parent is not aware of constraints defined in other models.

I suggested we add some kind of warning for such cases. Fixing that is not possible on a stable version.
